### PR TITLE
Provider service factory model manager

### PR DESCRIPTION
**Pull request by SimonRichardson**
This updates the model environ tracker to use the database instead of going via the API connection to the apiserver. This requires the creation of a provider service. The provider service is a lightweight service factory that has no dependencies on the legacy state package.

The environ tracker has been renamed to provider tracker. It has been re-written to use change stream watchers directly. In doing so, we push the provider service factory through the modelmanager worker. As the modelmanager worker already depends on the service factory, it should be forced to restart if the is a problem with the database worker dependency. As an aside, we should convert as many modelmanager workers to use the database directly. This should remove a lot of load from the API server and could remove a lot of latency in the system. 

The same watch semantics for the provider tracker are employed as the prior environ tracker. Model config, cloud, and credential changes should trigger updates to the underlying environ.

These changes only take into consideration the provider (IAAS) and not the broker (CAAS). A follow-up PR will generalize the problem so one worker can update the provider and broker using the same code (generics). 

TODO:

 - [x] The provider bootstraps, but adding a new model fails because we don't create the model correctly in the modelmanager. A PR will need to be supplied prior to this set of patches to ensure that we do correctly write all the information. Sync with @tlm 

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap aws test
$ juju add-model default
$ juju update-credential aws --region eu-west-2
$ juju show-credential
```

## Links

**Jira card:** JUJU-5643

### CODE REVIEW

File: domain/state.go [hunk 1]
```diff
@@ -15,39 +15,50 @@ import (
 // StateBase defines a base struct for requesting a database. This will cache
 // the database for the lifetime of the struct.
 type StateBase struct {
-	mu    sync.Mutex
-	getDB database.TxnRunnerFactory
-	db    database.TxnRunner
+	dbMutex sync.RWMutex
```
Comment (manadart) [inline domain/state.go:18]: Could have left this out of the churn.

File: environs/config/config.go [hunk 2]
```diff
@@ -1994,425 +2003,3 @@ func developerConfigValue(name string) bool {
 	}
 	return false
 }
-
-// Schema returns a configuration schema that includes both
```
Comment (manadart) [inline environs/config/config.go:1998]: And this.

File: internal/worker/filenotifywatcher/watcher.go [hunk 3]
```diff
@@ -95,20 +96,31 @@ type Watcher struct {
...
 
+// NewWatcher returns a new FileWatcher that watches the given fileName in the
+// given path.
 func NewWatcher(fileName string, opts ...Option) (FileWatcher, error) {
 	o := newOption()
 	for _, opt := range opts {
 		opt(o)
 	}
 
+	// Ensure that we create the watch path.
```
Comment (manadart) [inline internal/worker/filenotifywatcher/watcher.go:107]: Can't see which commit explains this addition.

**Comment by SimonRichardson** [comment]
> I don't understand where we're going with this. When I sent out the email after we discussed the design, I noted the dependency chain as:

> DBACCESSOR -> ENVIRONTRACKER -> SERVICEFACTORY

You can't get a provider tracker from nowhere. Also, I'm a bit taken aback, as I have strong vivid conversations about how this all worked and how I planned to wire this up. You need services that provide the things that a provider tracker requires. These are the following:

 - Model (this is the read-only model)
 - Cloud
 - Credential
 - Model config

From those constituent parts, we can assemble a provider. If you're stating that I should just have one provider service to get all that information, then I think that's not ideal. By utilizing the restricted subset of each service, we can piggyback off the work that already exists.

A provider tracker doesn't require the whole service factory, and by releasing it of the service factory burden we can elevate the cyclic dependency on state. The provider service factory is threaded through the model worker manager, and is passed to the model dependency engine. I've lifted the service factory into the dependency engine so that we can then reuse the provide tracker later on at the controller level. _All_ of this work in the PR is the groundwork to be able to reuse the provider-tracker at the model and then allow us to reuse it at the controller level. By elevating ourselves of the API connection on the provider tracker we allow it to be reusable (again we discussed this).

In subsequent PRs, I will have a PR to generalize the provider tracker to then be a broker tracker (discussed here https://github.com/juju/juju/pull/17060). Once that PR has landed, we will then have a PR to make the generalized tracker mutli-model (adding a runner). We talked about having a config to restrict the runner to only one tracker at the model level. I didn't want to do it one PR, as I knew there was a lot of work to be done, by doing it incrementally we can see the work being broken down to achieve this. Unfortunately, there have been a few road blocks in the way (no model config, no models, agent test etc, etc).

The dependency graph then becomes:

```
                   ┌────────────────┐                  
                   │                │                  
                   │                │                  
                   │   DBACCESSOR   │                  
                   │                │                  
                   │                │                  
                   └───────┬────────┘                  
                           │                           
        ┌──────────────────┤                           
        │                  │                           
┌───────▼───────┐          │                           
│               │          │                           
│               │          │                           
│   PROVIDER    │          │                           
│   SERVICE     │          │                           
│   FACTORY     │          │                           
│               │          │                           
│               │          │        ┌─────────────────┐
└───────┬───────┘          │        │                 │
        │                  │        │                 │
        │                  │        │      OTHER      │
┌───────▼───────┐          │        │      DEPS       │
│               │          │        │                 │
│               │          │        │                 │
│   PROVIDER    │          │        └───────┬─────────┘
│   TRACKER     │          │                │          
│               │          │                │          
│               │          │                │          
└───────┬───────┘          │                │          
        │                  │                │          
        │                  │                │          
        └─────ENVIRON────┐ │ ┌──────────────┘          
                         │ │ │                         
                         │ │ │                         
                  ┌──────▼─▼─▼──┐                      
                  │             │                      
                  │             │                      
                  │   SERVICE   │                      
                  │   FACTORY   │                      
                  │             │                      
                  │             │                      
                  └──────┬──────┘                      
                         │                             
                         │                             
                  ┌──────▼──────┐                      
                  │             │                      
                  │             │                      
                  │   STATE     │                      
                  │             │                      
                  │             │                      
                  └─────────────┘                      
```

We can then freely inject the multi-model provider tracker into service factory, as the only dependency it has on the dependency engine. Thus we remove all circular dependencies within the graph. 

> The environ tracker was multi in the same way as the DB accessor.

> The idea was that you could then get an environ direct from the provider service factory. But here we've got a provider service > factory that doesn't do that - it seems to be just a subset of the model service factory.

> How do we get an environ in the bootstrap worker with this mechanism?

File: apiserver/facades/client/modelmanager/services.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+type serviceFactoryGetter struct {
+	ctx facade.MultiModelContext
+}
+
+func (s serviceFactoryGetter) ServiceFactoryForModel(uuid coremodel.UUID) ServiceFactory {
+	return serviceFactory{serviceFactory: s.ctx.ServiceFactoryForModel(uuid)}
+}
+
+type serviceFactory struct {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/modelmanager/services.go:114]: Invariant interfaces. If we want to use interfaces at the call site, then these are the consequences.

File: cmd/jujud-controller/agent/machine/manifolds.go [hunk 5]
```diff
@@ -697,16 +697,18 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
...
+			StateName:                       stateName,
+			LogSinkName:                     logSinkName,
+			ServiceFactoryName:              serviceFactoryName,
+			ProviderServiceFactoryName:      providerServiceFactoryName,
+			MuxName:                         httpServerArgsName,
+			NewWorker:                       modelworkermanager.New,
+			NewModelWorker:                  config.NewModelWorker,
+			ModelMetrics:                    config.DependencyEngineMetrics,
+			Logger:                          loggo.GetLogger("juju.workers.modelworkermanager"),
+			GetProviderServiceFactoryGetter: modelworkermanager.GetProviderServiceFactoryGetter,
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud-controller/agent/machine/manifolds.go:711]: Not a fan, but convention :man_shrugging:

File: cmd/jujud-controller/agent/model/manifolds.go [hunk 6]
```diff
@@ -153,6 +157,16 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
 			Logger:        config.LoggingContext.GetLogger("juju.worker.apicaller"),
 		}),
 
+		// The provider service factory is used to access the provider service.
+		// It's injected into the model worker manager so that it can be used
+		// by the provider and broker workers.
+		providerServiceFactoryName: dependency.Manifold{
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud-controller/agent/model/manifolds.go:163]: Not if I want to reuse the __same__ worker in the controller manifold.

File: domain/state.go (hunk 1 above)
Comment (SimonRichardson) [reply domain/state.go:18]: I had a conversation about this the other day. I asked if I should add it, you said yes, and here it is.

File: internal/worker/filenotifywatcher/watcher.go (hunk 3 above)
Comment (SimonRichardson) [reply internal/worker/filenotifywatcher/watcher.go:107]: This was creating noise in the agent test suite, which made it even more than infuriating to diagnose what the underlying problem was.

File: cmd/jujud-controller/agent/machine_legacy_test.go [hunk 7]
```diff
@@ -73,6 +73,13 @@ import (
 //
 // Do not edit them to make the sync point work better. They're legacy and
 // should be treated as such, until we cut them over.
+//
+// Addendum: These tests also only work on the controller database. Now that
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud-controller/agent/machine_legacy_test.go:77]: As discussed today, this is here to stay until we get our integration tests, then the legacy tests disappear. I do wonder how much of the prime agent goes away also.

File: internal/worker/modelworkermanager/manifold.go [hunk 8]
```diff
@@ -27,19 +28,44 @@ type Logger interface {
...
+	// model upon request.
+	ProviderServiceFactoriesName string
+	// LogSinkName is the name of the corelogger.ModelLogger dependency.
+	LogSinkName string
+
+	// GetProviderServiceFactoryGetter is used to get the provider service
+	// factory getter from the dependency engine. This makes testing a lot
+	// simpler, as we can expose the interface directly, without the
+	// intermediary type.
+	GetProviderServiceFactoryGetter GetProviderServiceFactoryGetterFunc
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply internal/worker/modelworkermanager/manifold.go:59]: I need to think about this. The location of this isn't the common case, in reality the common case will be at the machine and service factory level, which requires access of an provider tracker for a given model uuid. And potentially a model that you don't own.

**Review by manadart — APPROVED** [review]
Let's move this forward.

**Pull request merged**